lib/middleware-common/src/metering.rs

```diff
@@ -1,37 +1,118 @@
 use wasmer_runtime_core::{
-    codegen::{Event, EventSink, FunctionMiddleware, InternalEvent},
+    codegen::{Event, EventSink, FunctionMiddleware, InternalEvent::*},
     module::ModuleInfo,
     vm::{Ctx, InternalField},
     wasmparser::{Operator, Type as WpType, TypeOrFuncType as WpTypeOrFuncType},
     Instance,
 };
 
-static INTERNAL_FIELD: InternalField = InternalField::allocate();
+static INTERNAL_FIELD_USED: InternalField = InternalField::allocate();
+static INTERNAL_FIELD_LIMIT: InternalField = InternalField::allocate();
 
-/// Metering is a compiler middleware that calculates the cost of WebAssembly instructions at compile
-/// time and will count the cost of executed instructions at runtime. Within the Metering functionality,
-/// this instruction cost is called `points`.
+/// Metering is a compiler middleware that calculates the cost of WebAssembly instructions at
+/// compile time and will count the cost of executed instructions at runtime. Within the Metering
+/// functionality, this instruction cost is called `points`.
 ///
-/// The Metering struct takes a `limit` parameter which is the maximum number of points which can be
-/// used by an instance during a function call. If this limit is exceeded, the function call will
-/// trap. Each instance has a `points_used` field which can be used to track points used during
-/// a function call and should be set back to zero after a function call.
+/// Each instance has an `exec_limit` which is the maximum number of points which can be used by
+/// the instance during a function call. If this limit is exceeded, the function call will trap.
+/// Each instance has a `points_used` field which can be used to track points used during a
+/// function call and should be set back to zero after a function call.
 ///
 /// Each compiler backend with Metering enabled should produce the same cost used at runtime for
 /// the same function calls so we can say that the metering is deterministic.
-///
 pub struct Metering {
-    limit: u64,
-    current_block: u64,
+    injections: Vec<Injection>,
+    current_block_injections: Vec<Injection>,
+    current_block_cost: u64,
 }
 
 impl Metering {
-    pub fn new(limit: u64) -> Metering {
+    pub fn new() -> Metering {
         Metering {
```

Comment:
Small improvement, but you can use the keyword `Self` in `impl X` contexts to refer to the type being created, so this can be 

```Rust
pub fn new() -> Self {
    Self { ... }
}
```

One more improvement, is that we can add `#[derive(Default)]` to the struct (should look like `#[derive(Debug, Default)]`) and then do something like:

```Rust
pub fn new() -> Self {
   Self::default()
}
```

These are very minor points, but it makes editing the code a bit quicker by reducing redundancy; and if in the future we want a default struct but with some number of fields not default, we can use the syntax:

```Rust
Self {
   override_field1: value1,
   override_field2: value2,
   ..Self::default()
}
```

Reply:
Thanks! I really appreciate the suggestions so I can improve my Rust skills.